### PR TITLE
[mempool] Fix upstream peers counter

### PR DESCRIPTION
**Pull request by gregnazario**
Here's a silly bug, where we double count an upstream peer if we received a new connection, but it was already in our records as alive.

### CODE REVIEW

File: mempool/src/shared_mempool/peer_manager.rs [hunk 1]
```diff
@@ -118,11 +118,14 @@ impl PeerManager {
         let mut peer_states = self.peer_states.lock();
         let is_new_peer = !peer_states.contains_key(&peer);
         if self.is_upstream_peer(&peer, Some(&metadata)) {
-            counters::active_upstream_peers(&peer.raw_network_id()).inc();
             // If we have a new peer, let's insert new data, otherwise, let's just update the current state
             if is_new_peer {
+                counters::active_upstream_peers(&peer.raw_network_id()).inc();
                 peer_states.insert(peer, PeerSyncState::new(metadata));
             } else if let Some(peer_state) = peer_states.get_mut(&peer) {
+                if !peer_state.is_alive {
```
Comment (JoshLind) [inline mempool/src/shared_mempool/peer_manager.rs:126]: Shouldn't we decrement the alive peers if we're still going to add them again (when they are alive)? If not, we can still get double counts, right? i.e., add peer x (active peers = 1), remove peer X (which disables it, active peers = 1), add peer x (which enables it here, but re-increments, which means active peers = 2, but it's still only 1 peer?)

Comment (gregnazario) [reply mempool/src/shared_mempool/peer_manager.rs:126]: Yeah, I realize now it never decremented.  Fixed now

File: mempool/src/shared_mempool/peer_manager.rs [hunk 2]
```diff
@@ -141,12 +144,14 @@ impl PeerManager {
...
         if peer.raw_network_id().is_validator_network() {
             if let Some(state) = self.peer_states.lock().get_mut(&peer) {
+                counters::active_upstream_peers(&peer.raw_network_id()).dec();
                 state.is_alive = false;
             }
         } else {
             // All other nodes have their state immediately restarted anyways, so let's free them
             // TODO: Why is the Validator optimization not applied here
             self.peer_states.lock().remove(&peer);
+            counters::active_upstream_peers(&peer.raw_network_id()).dec();
```
Comment (JoshLind) [inline mempool/src/shared_mempool/peer_manager.rs:154]: nit: I still think this logic is a little weird (i.e., as called out by the TODO). For example, I'd only decrement the counter if it's not in the peer_states (even in this case), to protect against any weird bugs where we get a removal notification for a peer but we never got a new peer notice, but this shouldn't happen. Either way, we should probably clean all this stuff up at some point 😄  For now, it's probably good enough!

**Review by JoshLind — APPROVED** [review]
Thanks @gregnazario! 😄

File: mempool/src/shared_mempool/peer_manager.rs (hunk 2 above)
Comment (gregnazario) [reply mempool/src/shared_mempool/peer_manager.rs:154]: Yeah, I don't get it either, but I didn't have time to figure out why it does it really...

**Pull request closed without merging**